### PR TITLE
BUG, Benchmark: fix passing optimization build options to asv

### LINKED ISSUE #17716: SIMD: The --cpu-baseline is not correctly configured in benchmark system.
I'm trying to use the newest benchmark system to test the performance, but found out that the option is not working. here is the command that used for benchmarking.
```
python runtests.py --cpu-baseline="avx2" --bench-compare master bench_linalg.Einsum.time_einsum_multiply
```
in this case, the macro `NPY_SIMD `should be 256 because my machine support `avx2 `intrinsics, but the actual value is 128, which is the default minimal settings.

### PR DESCRIPTION
**Pull request by seiko2plus**
#### Fix passing optimization build options to asv
closes #17716

see https://github.com/numpy/numpy/issues/17716#issuecomment-723712282

### CODE REVIEW

**Comment by Qiyu8** [comment]
I tested your code in two strategies:

-  single branch benchmark: create new branch, download your code, run the following command.

     ```python runtests.py --cpu-baseline="sse2" --bench bench_linalg```

   - [x] The `NPY_SIMD `is 128.

    ```python runtests.py --cpu-baseline="avx2" --bench bench_linalg```

   - [x] The `NPY_SIMD `is 256.

- two branch benchmark: create two new branch, download your code in both branch, run  the following command.

    ```python runtests.py --cpu-baseline="sse2" --bench-compare test_asv bench_linalg```
   - [x] The `NPY_SIMD `is 128.

    ```python runtests.py --cpu-baseline="avx2" --bench-compare test_asv bench_linalg```

   - [x] The `NPY_SIMD `is 256.

To sum it up, the code here works fine in my computer. Thanks @seiko2plus .

**Comment by seiko2plus** [comment]
@Qiyu8, thank you for reporting, you can also count on #17737 to verify the supported features. it provides the same format that used within pytest.

File: benchmarks/asv_compare.conf.json.tpl [hunk 1]
```diff
@@ -78,7 +78,9 @@
 
     "build_command" : [
         "python setup.py build {numpy_build_options}",
-        "PIP_NO_BUILD_ISOLATION=false python -mpip wheel --no-deps --no-index -w {build_cache_dir} {build_dir}"
+        // pip ignores '--global-option' when pep517 is enabled, we also enabling pip verbose to
+        // be reached from asv `--verbose` so we can verify the build options.
+        "PIP_NO_BUILD_ISOLATION=false python {build_dir}/benchmarks/asv_pip_nopep517.py -v {numpy_global_options} --no-deps --no-index -w {build_cache_dir} {build_dir}"
```
Comment (mattip) [inline benchmarks/asv_compare.conf.json.tpl:83]: This should be broken into two steps: one to check that `--no-use-pep517` is supported, and one to build the wheel. It might be easier to avoid the pip build altogether by adding a `bdist_wheel` build step
```
rm -rf dist
python setup.py build {numpy_build_options} bdist_wheel
python -mpip install dist/*.whl
```

Comment (seiko2plus) [reply benchmarks/asv_compare.conf.json.tpl:83]: > This should be broken into two steps: one to check that --no-use-pep517 is supported, and one to build the wheel

already handle it through a separate python script `asv_pip_nopep517.py`, we're not free to use POSIX commands since we're targeting non-unix platforms too. e.g. windows

> It might be easier to avoid the pip build altogether by adding a bdist_wheel build step

I tried to use `bdist_wheel` but ASV wouldn't able to fetch the wheel file and cache it.
also, I tried to only use the `pip` command only but idk why `ccache` doesn't work properly

here is the default values of build commands :
```Json
"install_command":
    ["python -mpip install {wheel_file}"],
"uninstall_command":
    ["return-code=any python -mpip uninstall -y {project}"],
"build_command":
    ["python setup.py build",
     "PIP_NO_BUILD_ISOLATION=false python -mpip wheel --no-deps --no-index -w {build_cache_dir} {build_dir}"]
```

**Comment by Qiyu8** [comment]
I think this PR is good to merge and will unlock other performance optimizations.

**Comment by mattip** [comment]
While I do not debate the value in passing optimization flags to asv, I think this PR still is not ready. There should be a simpler and more direct way to get interaction between asv and code building. 

Here are some ideas, it may be worth discussing them more widely:
- overhaul runtests.py and refactor it into proper subcommands rather than using home-grown argparse enhancements. This would allow refactoring the code, which might clarify the interactions between building NumPy and running benchmarks, and reveal a better path for this kind of code.
- enhance ASV and/or the interaction between the benchmarks, ASV, and NumPy. PRs gh-17737,  gh-15987 and gh-15992 are also related to this.

**Comment by seiko2plus** [comment]
@mattip,

> There should be a simpler and more direct way to get interaction between asv and code building.

unfortunately `asv continuous` doesn't offer any other way, check airspeed-velocity/asv#803

> Here are some ideas, it may be worth discussing them more widely

First, we need to take a decision to either accepting this pull request as a workaround or close it and revert #17726 

>  overhaul runtests.py and refactor it into proper subcommands rather than using home-grown argparse enhancements.

as a start, we can create an independent script `benchmark.py` that mapping all important args to ASV, it should
provides similar functionality to gh-15987

> enhance ASV and/or the interaction between the benchmarks, ASV, and NumPy. PRs gh-17737, gh-15987 and gh-15992 are also related to this.

Here is the issue, ASV is based on command arguments that kill the flexibility.
I vote to create our own micro-benchmarking tool and drop ASV.

**Comment by Qiyu8** [comment]
I think It's hard time to drop `asv `support, because this benchmark toolkit is working well among normal modules,  The current `arg `based simd mechanism exposed the drawbacks of `asv`, which should improved through [community](https://github.com/airspeed-velocity/asv).

**Comment by seiko2plus** [comment]
@Qiyu8, with all respect, we can't count on another project or community in every single new feature. I see [google benchmark](https://github.com/google/benchmark) as a good alternative, we can build our new tool on top of it. It already provides decent flexibility and been used by many projects.

**Comment by seiko2plus** [comment]
@mattip, there's no other way to get ride of it without this workaround. ugly but necessary and the same thing for #17737.
please I need to hear your final call.

**Comment by mattip** [comment]
I marked this for triage-review so we can get some more opinions.

**Comment by seiko2plus** [comment]
@mattip, thank you Matti, I really appreciate your help.

**Comment by mattip** [comment]
The SIMD optimizations is stretching the current workflows:
- a need to benchmark ufunc inner loops
- a new templating engine for C code
- a new compiler option extension for setting machine options to C flags (merged)

Maybe we should think a bit about separating the loops out into a separate library that would be easier to test and benchmark, along the lines of issue gh-17698.

**Comment by mattip** [comment]
In the triage meeting we decided to merge this, but would like to rethink benchmarking in general. That is a different discussion though, so in this goes.

**Comment by mattip** [comment]
Thanks @seiko2plus

**Comment by seiko2plus** [comment]
@mattip,

Infrastructure, SIMD, and GPU optimizations must be part of the core of any modern computing library, branching the core or counting on other projects will badly affect workflow, especially for big projects.

We have something new here, let's give it the full chance before thinking of separating it into a separate repo. I think NumPy should find a foot in highly computing processing instead of playing the mediator role, in order to maintain its popularity. 


> need to benchmark ufunc inner loops

We need also a tool for sanity data check too. OpenCV has 2 in 1 tool benchmarking, and sanity checks at the same time. see [HowToUsePerfTests](https://github.com/opencv/opencv/wiki/HowToUsePerfTests) very effective in detecting any deviations.

> a new templating engine for C code

Why do we need a new engine and we have python? we just need a better way to using it in our C sources. I think #17952 can be a good solution.

> In the triage meeting we decided to merge this,

Thank you a lot.

> but would like to rethink benchmarking in general.

I'm not sure what we should do. To create our own minimal tool that serving our only needs or contributing to ASV. What do you prefer?

**Comment by mattip** [comment]
In general, I do not like re-inventing the wheel. If a tool or paradigm can be adapted to our needs, then that is preferable over creating our own, unless our needs are very different.

> I think NumPy should find a foot in highly computing processing instead of playing the mediator role, in order to maintain its popularity.

I disagree. If by splitting out the great work you have done with SIMD we enable others to reuse it, that would be progress.

> What do you prefer?

It seems like right now ASV and pyperf are very similar, but neither serves our needs very well. I think we should explore ways to make ASV better before creating our own tool. Making those tools better would help others who are also struggling with benchmarking.

**Comment by rossbar** [comment]
> It seems like right now ASV and pyperf are very similar, but neither serves our needs very well. I think we should explore ways to make ASV better before creating our own tool. Making those tools better would help others who are also struggling with benchmarking.

Well said, I very much agree :+1:

**Comment by seiko2plus** [comment]
@mattip, I will give ASV a try I guess we have enough burden and no need to open a new front.
